arrows/core/video_input_arf.cxx

```diff
@@ -0,0 +1,558 @@
+  {
+    return false;
+  }
+
+  d->frame_number = frame_number;
+
+  // Return timestamp
+  ts = this->frame_timestamp();
+
+  return true;
+}
+
+// ------------------------------------------------------------------
+kv::timestamp
+video_input_arf
+::frame_timestamp() const
+{
+  // Check for at end of video
+  if ( this->end_of_video() )
+  {
+    return {};
+  }
+
+  kv::timestamp ts;
+
+  ts.set_frame( d->frame_number );
+
+  return ts;
+}
+
+// ------------------------------------------------------------------
+kv::image_container_sptr
+video_input_arf
+::frame_image()
+{
+  if (!good())
+    return nullptr;
+
+  auto img_bytes = std::make_shared<kv::image_memory>(d->img_size);
+
+  // TODO Questions if using the GAP_SIZE is correct or not..
+  fseek(d->file, d->offset + (d->frame_number * (d->img_size + GAP_SIZE)), SEEK_SET);
+  fread(img_bytes->data(), 1, d->img_size, d->file);
+  if (d->convert_endian)
+    d->byteswap_img_bytes((unsigned char*)img_bytes->data());
+
+  kv::image img(img_bytes,img_bytes->data(),
+    d->cols, d->rows, 1, 1, d->cols, 1, d->px);
+
+  auto m = std::make_shared<kv::metadata>();
+  m->set_timestamp(this->frame_timestamp());
+  auto imc = std::make_shared<kv::simple_image_container>(img);
+  imc->set_metadata(m);
+
+  return imc;
+}
+
+// ------------------------------------------------------------------
+kv::metadata_vector
+video_input_arf
+::frame_metadata()
+{
+  kv::metadata_vector vect;
+  if ( d->metadata )
+  {
+    vect.push_back( d->metadata );
+  }
+
+  return vect;
+}
+
+// ----------------------------------------------------------------------------
+kv::metadata_map_sptr
+video_input_arf
+::metadata_map()
+{
+  return std::make_shared< kv::simple_metadata_map >(d->metadata_map);
+}
+
+// ----------------------------------------------------------------------------
+unsigned short
+video_input_arf::priv
+::byteswap(unsigned short value)
+{
+#ifdef WIN32
+  return _byteswap_ushort(value);
+#else
+  return __builtin_bswap16(value);
+#endif
+}
+
+// ----------------------------------------------------------------------------
+unsigned int
+video_input_arf::priv
+::byteswap(unsigned int value)
+{
+#ifdef WIN32
+  return _byteswap_ulong(value);
+#else
+  return __builtin_bswap32(value);
+#endif
+}
+
+// ----------------------------------------------------------------------------
+unsigned long
+video_input_arf::priv
+::byteswap(unsigned long value)
+{
+#ifdef WIN32
+  return _byteswap_ulong(value);
+#else
+  return __builtin_bswap32(value);
+#endif
+}
+
+// ----------------------------------------------------------------------------
+unsigned __int64
+video_input_arf::priv
+::byteswap(unsigned __int64 value)
+{
+#ifdef WIN32
+  return _byteswap_uint64(value);
+#else
+  return __builtin_bswap64(value);
+#endif
+}
+
+// ----------------------------------------------------------------------------
+void
+video_input_arf::priv
+::byteswap_img_bytes(unsigned char* img_bytes)
```

Comment:
I'm not sure if it's worth rewriting this (maybe later?), but I can't help but wonder if rewriting this to iterate over a `rows*cols` integer array using `ntohX` wouldn't be faster. (It generates radically different, and shorter, assembly, at least...)